crates/sui-bcs-json/src/lib.rs

```diff
@@ -146,75 +147,143 @@ fn to_bcs<T: Serialize + for<'a> Deserialize<'a>>(data: &str) -> Result<Vec<u8>>
+        typename if typename.starts_with("vector") => {
+            MyValue::simple_deserialize(data, &TypeLayout::build(typename, 0)?)
+        }
+        _ => todo!(),
     }
 }
 
-fn to_json_from_vector(typename: &str, data: &mut &[u8]) -> Result<Value> {
-    let length = read_uleb128_as_u64(data)?;
-    let mut output = Vec::with_capacity(length as usize);
-    for _ in 0..length {
-        output.push(internal_bcs_to_json_object(typename, data)?);
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum MyValue {
+    U8(u8),
+    U16(u16),
+    U32(u32),
+    U64(u64),
+    U128(u128),
+    U256(U256),
+    Bool(bool),
+    Vector(Vec<MyValue>),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum TypeLayout {
+    Bool,
+    U8,
+    U16,
+    U32,
+    U64,
+    U128,
+    U256,
+    Vector(Box<TypeLayout>),
+}
+
+impl TypeLayout {
+    fn build(t: &str, depth: u64) -> Result<TypeLayout> {
+        Ok(match t {
+            "bool" => TypeLayout::Bool,
+            "u8" => TypeLayout::U8,
+            "u16" => TypeLayout::U16,
+            "u32" => TypeLayout::U32,
+            "u64" => TypeLayout::U64,
+            "u128" => TypeLayout::U128,
+            "u256" => TypeLayout::U256,
+            typename if t.starts_with("vector") => {
+                let inner_type = is_vector(typename).unwrap();
+                TypeLayout::Vector(Box::new(Self::build(inner_type, depth + 1)?))
+            }
+            _ => todo!(),
+        })
     }
-    Ok(output.into())
 }
 
-fn internal_bcs_to_json(typename: &str, data: &mut &[u8], pretty: bool) -> Result<String> {
-    println!("bcs_to_json: {:?}", typename);
-    match internal_bcs_to_json_object(typename, data) {
-        Ok(value) if pretty => Ok(serde_json::to_string_pretty(&value)?),
-        Ok(value) => Ok(serde_json::to_string(&value)?),
-        Err(e) => bail!("Cannot convert bcs to json string: {e}"),
+impl MyValue {
+    pub fn simple_deserialize(blob: &[u8], ty: &TypeLayout) -> Result<Value> {
+        Ok(serde_json::to_value(bcs::from_bytes_seed(ty, blob)?)?)
+    }
+    pub fn simple_serialize(&self) -> Option<Vec<u8>> {
+        bcs::to_bytes(self).ok()
     }
 }
 
-/// Deserializes a BCS array to an integer.
-///
-/// By default, u8, u16, u32 are serialized as JSON numbers,
-/// while u64, u128, and U256 are serialized as JSON strings.
-fn string_integer_to_json(typename: &str, data: &mut &[u8]) -> Result<Value> {
-    match typename {
-        "u8" => {
-            let bcs: u8 = bcs::from_bytes(data)?;
-            Ok(serde_json::to_value(bcs)?)
-        }
-        "u16" => {
-            let bcs: u16 = bcs::from_bytes(data)?;
-            Ok(serde_json::to_value(bcs)?)
-        }
-        "u32" => {
-            let bcs: u32 = bcs::from_bytes(data)?;
-            Ok(serde_json::to_value(bcs)?)
-        }
-        "u64" => {
-            let bcs: u64 = bcs::from_bytes(data)?;
-            Ok(serde_json::to_value(bcs.to_string())?)
+impl serde::Serialize for MyValue {
```

Comment:
The compiler complained that there is no impl of `Serialize` for `MyValue`, thus I added one.